gdnative-core/src/export/method.rs

```diff
@@ -284,16 +287,410 @@ impl<'a> Varargs<'a> {
+    #[inline]
+    #[must_use]
+    fn ordinal(&self) -> String {
+        match self.index + 1 {
+            1 => "1st".to_owned(),
+            2 => "2nd".to_owned(),
+            3 => "3rd".to_owned(),
+            i @ 4.. => format!("{i}th"),
+            _ => "unknown".to_owned(),
+        }
+    }
+
+    /// Get the argument type error's index.
+    #[inline]
+    #[must_use]
+    pub fn index(&self) -> usize {
+        self.index
+    }
+
+    /// Get a reference to the argument type error's nested error.
+    #[inline]
+    #[must_use]
+    pub fn nested_error(&self) -> &FromVariantError {
+        &self.nested_error
+    }
+}
+
+impl std::error::Error for ArgumentTypeError {}
+impl std::fmt::Display for ArgumentTypeError {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Incorrect type of {} argument, Cause: {}",
+            self.ordinal(),
+            self.nested_error,
+        )
+    }
+}
+
+/// Error to argument lengths do not match.
+#[derive(Debug)]
+pub struct ArgumentLengthError {
+    passed: usize,
+    expected_min: usize,
+    expected_max: usize,
+}
+
+impl ArgumentLengthError {
+    /// Creates a new error with the length of the arguments passed and the expected arguments range.
+    #[inline]
+    #[must_use]
+    pub fn new(passed: usize, expected: impl ArgumentBounds) -> Self {
+        Self {
+            passed,
+            expected_min: expected.start_bound(),
+            expected_max: expected.end_bound(),
+        }
+    }
+
+    /// Get the argument length error's passed.
+    #[inline]
+    #[must_use]
+    pub fn passed(&self) -> usize {
+        self.passed
+    }
+
+    /// Get the argument length error's expected min.
+    #[inline]
+    #[must_use]
+    pub fn expected_min(&self) -> usize {
+        self.expected_min
+    }
+
+    /// Get the argument length error's expected max.
+    #[inline]
+    #[must_use]
+    pub fn expected_max(&self) -> usize {
+        self.expected_max
+    }
+}
+
+impl std::error::Error for ArgumentLengthError {}
+impl std::fmt::Display for ArgumentLengthError {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let expected_msg = match (self.expected_min, self.expected_max) {
+            (usize::MIN, usize::MAX) => "any".to_owned(),
+            (usize::MIN, e) => format!("max {e}"),
+            (s, usize::MAX) => format!("min {s}"),
+            (s, e) => {
+                if s == e {
+                    s.to_string()
+                } else {
+                    format!("min {s} and max {e}")
+                }
+            }
+        };
+        write!(
+            f,
+            "Argument lengths do not match, passed {}, expected {}",
+            self.passed, expected_msg
+        )
+    }
+}
+
+/// Produced by range syntax like `a`, `(a, b)`, ..`, `a..`, `..b`, `..=c`, `d..e`, or `f..=g`.
+pub trait ArgumentBounds {
```

Comment:
Could we maybe reuse something like [`std::slice::SliceIndex`](https://doc.rust-lang.org/std/slice/trait.SliceIndex.html) instead?

Would be good to balance the amount of customization for rarely-used code, even if it means slightly reduced ergonomics in some cases.

Reply:
No. It cannot be used to validate argument lengths. This is because only methods are defined to retrieve slices.

Reply:
I see. What I don't like however is that we're adding hundreds of lines of code for shiny syntax in a rarely used corner of the library. The whole `ArgumentBounds` type has one single use case:
```rs
args.check_length(1..=3)
//                ^^^^^
``` 
While it's admittedly a very nice syntax, most users don't even get in contact with `Varargs` at all. The issue is that more code means more tests, higher complexity, more risk of bugs, longer compile times, larger maintenance burden. This sounds exaggerated for the case here, but these things do add up, and as the only maintainer of godot-rust who is active on an almost daily basis, I tend to be a skeptic in this regard 🙂

Reply:
I really think we could keep this whole functionality much simpler. We don't need `ArgumentBounds` at all and can use [`std::ops::RangeBounds`](https://doc.rust-lang.org/std/ops/trait.RangeBounds.html) instead:
```rs
impl<'a> Varargs<'a> {
    #[inline]
    pub fn check_length(&self, bounds: impl RangeBounds<usize>) -> Result<(), ArgumentLengthError> {
        let passed = self.args.len();
        if bounds.contains(&passed) {
            Ok(())
        } else {
            Err(ArgumentLengthError::new(passed, bounds))
        }
    }
}
```
Then, the whole `ArgumentLengthError` could also be kept simple. Initially we don't need to expose getters, a message will be fine. If people truly need access to the details, we can always add getters later.
```rs
struct ArgumentLengthError {
    message: String,
}

impl ArgumentLengthError {
    fn new(len: usize, bounds: impl RangeBounds<usize>) -> Self {
        Self {
            message: format!(
                "Argument length {} does not match expected bounds: {:?} to {:?}",
                len,
                bounds.start_bound(),
                bounds.end_bound()
            ),
        }
    }
}

impl std::fmt::Debug for ArgumentLengthError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Display::fmt(self, f)
    }
}

impl std::fmt::Display for ArgumentLengthError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for ArgumentLengthError {}
```

The good thing is, it still allows this:
```rs
fn main() {
    let s = S { args: vec![(); 6] };
    s.check_length(3..5).expect("validation");
}
```
with runtime error:
```
thread 'main' panicked at 'validation: Argument length 6 does not match expected bounds: Included(3) to Excluded(5)', src\main.rs:53:26
```
which is maybe not 100% as shiny, but definitely understandable enough, and needs only a fraction of the code.

Reply:
Understood. I' ll change it to use `RangeBounds<usize>`. I was too eager. 😓